### PR TITLE
Fix format_number NaN symbol in high jdk version

### PR DESCRIPTION
**Pull request by thirtiseven**
Fix https://github.com/NVIDIA/spark-rapids/issues/10175

In format_number, the symbol representing NaN is from `DecimalFormatSymbols.getInstance(Locale.US).getNaN` in Spark, but is hardcoded in the kernel in the plugin. So it will mismatch for NaN in different JDK versions.

This PR will check the NaN symbol when running format_number, if it is different from the one we use in the kernel, then find the NaNs in the results and replace them with the correct symbol. 

Tested locally with jdk 11 and verified that the code doesn't go into the replace nan branch under jdk 8.

perf test results:

| Data Type | jdk 11 GPU Time (ms) | jdk 8 GPU Time (ms) | jdk 11 CPU Time (ms) | jdk 11 Speed up |
|-|-|-|-|-|
| double | 633 | 321 | 10641 | 16.81x |
| float | 247 | 230 | 4602 |  18.63x |

### CODE REVIEW

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala [hunk 1]
```diff
@@ -2296,7 +2297,20 @@ case class GpuFormatNumber(x: Expression, d: Expression)
     val d = rhs.getValue.asInstanceOf[Int]
     x.dataType match {
       case FloatType | DoubleType => {
-        CastStrings.fromFloatWithFormat(lhs.getBase, d)
+        val nanSymbol = DecimalFormatSymbols.getInstance(Locale.US).getNaN
```
Comment (firestarman) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala:2300]: Do we need to fix this inside the `CastStrings.fromFloatWithFormat` ?
Other places calling this function may have the same issue ?

Comment (thirtiseven) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala:2300]: The only use case I know for this kernel is format_number, but fixing it in kernel would also prevent it from losing performance. I'm happy to fix it in the kernel if you think it's necessary now.

Comment (firestarman) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala:2300]: I don't know the whole backgroud. If this is only the place to call this function, the current change is fine with me.

**Comment by firestarman** [comment]
LGTM

**Comment by thirtiseven** [comment]
build

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala [hunk 2]
```diff
@@ -2296,7 +2297,20 @@ case class GpuFormatNumber(x: Expression, d: Expression)
...
+        // JDK 8's nan symbol is "�" ('\uFFFD'), which is also the jni kernel's nan symbol
+        // In higher JDK version, nan symbol is "NaN", so we need to handle it here.
```
Comment (jlowe) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala:2302]: We already have a custom kernel to handle formatting, and it's not hard to pass parameters to CUDA kernels.  Is there a plan to fix this directly in the kernel at some point?  Needing to do fixups after the custom kernel increases memory pressure and reduces performance, and we have the ability to update the custom kernel do the right thing from the start.

Comment (thirtiseven) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala:2302]: Yes, fix in the kernel is better, I plan to do it soon in a follow-up if you think ok. I chose a quicker fix in this PR to unblock JDK 11 & 17 nightly CI.

Comment (jlowe) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala:2302]: Yes that's totally fine, just checking that this isn't considered the final fix.  Thanks!

**Review by jlowe — APPROVED** [review]
@thirtiseven Please file the followup issue to fix this in the kernel and remove the patch-up workaround in the plugin.

**Comment by thirtiseven** [comment]
followup issue: https://github.com/NVIDIA/spark-rapids/issues/10183

**Pull request merged**